src/linear_algebra/matrix/charpoly/linear_map.lean

```diff
@@ -0,0 +1,308 @@
+/-
+Copyright (c) 2022 Andrew Yang. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Andrew Yang
+-/
+import linear_algebra.matrix.charpoly.coeff
+import linear_algebra.matrix.to_lin
+
+/-!
+
+# Calyley-Hamilton theorem for arbitrary rings and f.g. modules.
+
+-/
+
+variables (ι : Type*) [decidable_eq ι]
+variables {M : Type*} [add_comm_group M] {R : Type*} [comm_ring R] [module R M] (I : ideal R)
+variables (v : ι → M) (hv : submodule.span R (set.range v) = ⊤)
+
+variables {M' : Type*} [add_comm_monoid M'] [module R M']
+variables (v' : ι → M') (hv' : submodule.span R (set.range v') = ⊤)
+
+-- def linear_map.fun_right (f : module.End R M') : (ι → M) →ₗ[R] (ι → M') :=
+-- { to_fun := λ x i, f (x i),
+--   map_add' := λ x y, funext $ λ i, f.map_add _ _,
+--   map_smul' := λ x y, funext $ λ i, f.map_smul _ _ }
+
+variables (M)
+
+open_locale big_operators
+
+noncomputable
+def ideal.finsupp_total : (ι →₀ I) →ₗ[R] M :=
+(finsupp.total ι M R v).comp (finsupp.map_range.linear_map I.subtype)
+
+variables {ι M v}
+
+@[simp] lemma ideal.finsupp_total_apply (f : ι →₀ I) :
+  ideal.finsupp_total ι M I v f = f.sum (λ i x, (x : R) • v i) :=
+begin
+  dsimp [ideal.finsupp_total],
+  rw [finsupp.total_apply, finsupp.sum_map_range_index],
+  exact λ _, zero_smul _ _
+end
+
+lemma ideal.finsupp_total_apply_eq_of_fintype [fintype ι] (f : ι →₀ I) :
+  ideal.finsupp_total ι M I v f = ∑ i, (f i : R) • v i :=
+begin
+  rw ideal.finsupp_total_apply,
+  apply finset.sum_subset (finset.subset_univ _),
+  intros x _ hx,
+  rw finsupp.not_mem_support_iff.mp hx,
+  exact zero_smul _ _
+end
+
+lemma ideal.range_finsupp_total :
+  (ideal.finsupp_total ι M I v).range = I • (submodule.span R (set.range v)) :=
+begin
+  apply le_antisymm,
+  { rintros x ⟨f, rfl⟩,
+    rw ideal.finsupp_total_apply,
+    apply sum_mem _, { apply_instance },
+    intros c _,
+    apply submodule.smul_mem_smul (f c).2,
+    apply submodule.subset_span,
+    exact set.mem_range_self c },
+  { rw submodule.smul_le,
+    rintros r hr m hm,
+    rw ← set.image_univ at hm,
+    obtain ⟨l, hl, rfl⟩ := (finsupp.mem_span_image_iff_total _).mp hm,
+    let l' : ι →₀ I := finsupp.map_range (λ x : R, (⟨x * r, I.mul_mem_left _ hr⟩ : I))
+      (subtype.ext $ zero_mul _) l,
+    use l',
+    rw [ideal.finsupp_total_apply, finsupp.total_apply, finsupp.sum, finsupp.sum, finset.smul_sum],
+    dsimp,
+    simp only [← mul_smul, mul_comm r],
+    apply finset.sum_subset,
+    { exact finsupp.support_map_range },
+    { intros x hx hx',
+      have : l x * r = 0 := by injection finsupp.not_mem_support_iff.mp hx',
+      rw [this, zero_smul] } }
+end
+
+variables (v) [fintype ι]
+
+variables (R M ι)
+
+noncomputable
+def fintype.total : (ι → R) →ₗ[R] M :=
```

Comment:
Isn't this a duplicate?

Reply:
Yeah. This is still a draft and I'm still moving things around.